### PR TITLE
Attest separately for each API key and environment

### PR DESCRIPTION
**Pull request by davidme-stripe**
## Summary
We don't share attestation keys across API keys, so we need a separate attestation key for each API key. This slightly increases the number of attestations, but prevents situations where switching between merchants causes attestations to fail until the hardcoded timeout.

We'll also relax the 24 hour limit when using the development App Attest environment, as it doesn't increment the device's risk metric. We will still enforce it for generating keys in the production environment. To determine if we're in the development environment, we'll look for the appattestdevelop aaguid in the attestation data blob.

## Motivation
Switching between merchants and environments without waiting for the 24 hour attestation failsafe to time out.

## Testing
Updated existing tests, added a new test, tested attestations on physical device.

## Changelog
None

### CODE REVIEW

File: StripeCore/StripeCore/Source/Attestation/StripeAttest.swift [hunk 1]
```diff
@@ -117,33 +117,42 @@ import UIKit
...
         }
         set {
-            UserDefaults.standard.set(newValue, forKey: DefaultsKeys.lastAttestedDate.rawValue)
+            UserDefaults.standard.set(newValue, forKey: defaultsKeyForSetting(.lastAttestedDate))
         }
     }
 
+    /// The key to use for storing an attestation key in NSUserDefaults.
+    func defaultsKeyForSetting(_ setting: SettingsKeys) -> String {
+        var key = "\(setting.rawValue):\(apiClient.publishableKey ?? "unknown")"
```
Comment (wooj-stripe) [inline StripeCore/StripeCore/Source/Attestation/StripeAttest.swift:149]: Storing "unknown:xxx" in NSUserDefaults feels awkward. If we don't have a publishable key, seems like we should just bail?

File: StripeCore/StripeCoreTests/Attestation/MockAppAttestService.swift [hunk 2]
```diff
@@ -66,10 +67,14 @@ class MockAppAttestService: AppAttestService {
...
         // Generate a fake attestion
-        let attestation = ["keyID": key.id, "counter": key.counter, "clientDataHash": clientDataHash.base64EncodedString()] as [String: Any]
+        let attestation = ["keyID": key.id, "counter": key.counter, "clientDataHash": clientDataHash.base64EncodedString(), "isDevelopmentEnvironment": attestationUsingDevelopmentEnvironment] as [String: Any]
         return try JSONSerialization.data(withJSONObject: attestation)
     }
 
+    @_spi(STP) public func attestationDataIsDevelopmentEnvironment(_ data: Data) -> Bool {
+        let decodedKey = try! JSONSerialization.jsonObject(with: data) as! [String: Any]
+        return decodedKey["isDevelopmentEnvironment"] as! Bool
+    }
```
Comment (wooj-stripe) [inline StripeCore/StripeCoreTests/Attestation/MockAppAttestService.swift:77]: Feels like we can avoid using `as!` and fail gracefully?  E.g. if we can't decode, default to development.

Comment (davidme-stripe) [reply StripeCore/StripeCoreTests/Attestation/MockAppAttestService.swift:77]: This is test code, so I want it to fail loudly

File: StripeCore/StripeCore/Source/Attestation/StripeAttest.swift (hunk 1 above)
Comment (davidme-stripe) [reply StripeCore/StripeCore/Source/Attestation/StripeAttest.swift:149]: Let's bail above in `getOrCreateKey()` (though I'll still safely unwrap here so we don't crash)

File: StripeCore/StripeCore/Source/Attestation/StripeAttest.swift [hunk 3]
```diff
@@ -117,33 +119,42 @@ import UIKit
         case successfullyAttested = "STPAttestKeySuccessfullyAttested"
     }
 
-    private static var keyID: String? {
+    private var keyID: String? {
```
Comment (wooj-stripe) [inline StripeCore/StripeCore/Source/Attestation/StripeAttest.swift:122]: keyID is used way too many times in this file, can we rename this to something like `storedKeyID` ?

File: StripeCore/StripeCore/Source/Attestation/StripeAttest.swift (hunk 1 above)
Comment (wooj-stripe) [reply StripeCore/StripeCore/Source/Attestation/StripeAttest.swift:151]: Maybe we take a copy of apiClient.publishableKey when the class is initialized?

**Review by wooj-stripe — APPROVED** [review]

**Pull request merged**